src/librustc/hir/lowering.rs

```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_generic_arg(&mut self,
-                         arg: &ast::GenericArg,
-                         itctx: ImplTraitContext<'_>)
-                         -> hir::GenericArg {
+    fn lower_generic_arg(
+        &mut self,
+        arg: &ast::GenericArg,
+        itctx: ImplTraitContext<'_>
+    ) -> hir::GenericArg {
         match arg {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
+                                Res::Err | Res::Def(DefKind::Ctor(_, CtorKind::Fictive), ..) => (),
+                                res if res.is_valid_in_ns(Namespace::ValueNS) => {
+                                    debug!(
+                                        "lower_generic_arg: Lowering incorrectly parsed type \
+                                        argument as const argument: {:?}",
+                                        ty,
+                                    );
+
+                                    // Construct a AnonConst where the expr is the "ty"'s path.
+
+                                    let parent_def_index =
+                                        self.current_hir_id_owner.last().unwrap().0;
+                                    let node_id = self.sess.next_node_id();
+
+                                    // Add a definition for the in-band const def.
+                                    self.resolver.definitions().create_def_with_parent(
+                                        parent_def_index,
+                                        node_id,
+                                        DefPathData::AnonConst,
+                                        ExpnId::root(),
+                                        ty.span,
+                                    );
+
+                                    let path_expr = Expr {
+                                        id: ty.id,
+                                        kind: ExprKind::Path(qself.clone(), path.clone()),
```

Comment:
Can `qself` and `path` be consumed here instead of being cloned?

Reply:
The `GenericArg` we are matching against and getting these parameters from is not consumed, only passed as a reference, so sadly this clone needs to stay.